cmse/cmse.md

```diff
@@ -0,0 +1,1725 @@
+regular function call, as is the return from a non-secure function call. It is
+therefore required that a non-secure executable file can be developed using a
+toolchain that is not aware of CMSE.
+
+Developing a secure executable file requires toolchain support whenever a
+function is called from, calls, or returns to non-secure state and whenever
+memory is accessed through an address provided by the non-secure state. The
+secure code ABI is otherwise identical to the non-secure code ABI.
+
+For Armv8-M Mainline there are occasions where there is no possible
+code-generation that abides the VFP ABI rule that mandates that
+exception-control bits of the FPSCR may only be modified by specific support
+functions.  After returning from an entry call or when performing a
+nonsecure-call using the hard-float ABI the FPSCR will have been initialized
+with the secure world’s default FPSCR, the FPDSCR_S. For the soft-float ABI
+this is not the case as the vlstm and vldm instructions may be used. Armv8.1-M
+Mainline introduces instructions that enable the saving and restoring of the FP
+context. These are the VMSR, VMRS, VSTR and VLDR to system registers FPCXTNS and
+FPCXTS.
+
+The interaction between developers of secure code, non-secure code, and
+(optional) security agnostic library code is shown in [Figure 3](#figure3).
+
+![](figures/files-shared.png)
+
+<span id="figure3" class="citation-label">**Figure 3: Files shared between parties**</span>
+
+The secure gateway import library, shortened to _import library_, contains the
+addresses of the secure gateways of the secure code. This import library
+consists of or contains a relocatable file that defines symbols for all the
+secure gateways. The non-secure code links against this import library to use
+the functionality provided by the secure code.
+
+> 8	A relocatable file containing only copies of the (absolute) symbols of the
+> secure gateways in the secure executable must be available to link non-secure
+> code against.
+
+Linking against this import library is the only requirement on the toolchain
+used to develop the non-secure code. This functionality is very similar to
+calling ROM functions, and is expected to be available in existing toolchains.
+
+###	Secure gateway veneers
+
+> 9	A toolchain must support generating a _secure gateway veneer_ for each entry
+> function with external linkage. It consists of an `SG` instruction followed by
+> a `B.W` instruction that targets the entry function it veneers.
+
+Secure gateway veneers decouple the addresses of secure gateways (in NSC
+regions) from the rest of the secure code. By maintaining a vector of secure
+gateway veneers at a forever-fixed address, the rest of the secure code can be
+updated independently of non-secure code. This also limits the amount of code
+in NSC regions that potentially can be called by the non-secure state.
+
+> 10	A secure gateway veneer must be labelled by an ELF symbol that has the
```

Comment:
Here too: why use TAB? It gets rendered as a single space in the final docs anyway. You might want to use `&nbsp;` if you want to enforce multiple spaces?